### PR TITLE
Bugfix Json Schema, Improve json schema tests

### PR DESCRIPTION
**Pull request by simitt**
* Fix bugs uncovered with improved tests
* Do automatic validation for keyword limitations
* Remove files for invalid payloads, instead create invalid payloads out of valid payload
* Allow to define simple and conditional rules for testing payloads.

This PR is based on updating `github.com/santhosh-tekuri/jsonschema` https://github.com/elastic/apm-server/pull/1013 and on removing archived library `github.com/fatih/set` https://github.com/elastic/apm-server/pull/1014.

### CODE REVIEW

File: docs/spec/transactions/transaction.json [hunk 1]
```diff
@@ -50,8 +50,7 @@
             "regexProperties": true,
             "patternProperties": {
                 "^[^.*\"]*$": {
-                    "$ref": "mark.json",
-                    "maxLength": 1024
```
Comment (simitt) [inline docs/spec/transactions/transaction.json]: This length restriction did not make sense, as it doesn't refer to the properties key but the value and the value is a `mark`.

File: tests/json_schema.go [hunk 2]
```diff
@@ -11,13 +12,266 @@ import (
...
+	// than on intake API
+	TemplateToSchemaMapping map[string]string
+	LoadTemplatePaths       []string
+
+	TestData []TestData
+}
+
+type TestData struct {
+	Key       string
+	Invalid   []interface{}
```
Comment (graphaelli) [inline tests/json_schema.go]: Don't we lose the checks for why the data is invalid with this?  Not that comparing strings was great but it was actually useful

**Comment by simitt** [comment]
@graphaelli I'll add a rework to check for error messages and improve the `assertions`. I'll close the PR for now to avoid the noise, and reopen later.

**Comment by simitt** [comment]
I incorporated your suggestions @graphaelli:
- fail fast when assertion fails for precondition
- add expected error messages to validation failure checks
- add more comments

File: tests/json_schema.go [hunk 3]
```diff
@@ -27,7 +281,8 @@ type Mapping struct {
 }
 
 func TestPayloadAttributesInSchema(t *testing.T, name string, undocumentedAttrs *Set, schema string) {
-	payload, _ := loader.LoadValidData(name)
+	payload, err := loader.LoadValidData(name)
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (simitt) [reply tests/json_schema.go]: Missed that one - changing it now.

File: tests/json_schema.go [hunk 4]
```diff
@@ -1,23 +1,293 @@
...
+	TemplatePaths []string
+}
+
+type SchemaTestData struct {
+	Key       string
+	Valid     []interface{}
+	Invalid   Invalid
+	Condition Condition
+}
+type Invalid = map[string][]interface{}
```
[the remark this answers is not in the document]
Comment (simitt) [reply tests/json_schema.go]: good idea, changed

File: Makefile [hunk 5]
```diff
@@ -52,14 +52,14 @@ create-docs:
...
 	@mkdir -p docs/data/intake-api/generated/sourcemap
-	@cp tests/data/valid/error/payload.json docs/data/intake-api/generated/error/
-	@cp tests/data/valid/error/frontend.json docs/data/intake-api/generated/error/
-	@cp tests/data/valid/error/minimal_payload_exception.json docs/data/intake-api/generated/error/
-	@cp tests/data/valid/error/minimal_payload_log.json docs/data/intake-api/generated/error/
-	@cp tests/data/valid/transaction/payload.json docs/data/intake-api/generated/transaction/
-	@cp tests/data/valid/transaction/minimal_payload.json docs/data/intake-api/generated/transaction/
-	@cp tests/data/valid/transaction/minimal_span.json docs/data/intake-api/generated/transaction/
-	@cp tests/data/valid/sourcemap/bundle.js.map docs/data/intake-api/generated/sourcemap/
+	@cp tests/data/error/payload.json docs/data/intake-api/generated/error/
```
[the remark this answers is not in the document]
Comment (simitt) [reply Makefile]: I like the idea, will add it in an extra commit.

**Comment by simitt** [comment]
Thanks for the review @graphaelli. I liked your suggestions and pushed the changes.

**Comment by roncohen** [comment]
needs a backport to 6.x ?

**Pull request merged**